pkg/pillar/cmd/zedrouter/appnumonunet.go

```diff
@@ -0,0 +1,259 @@
+				appNumOnUNetKey(baseID, appID), appNum)
+			continue
+		}
+		log.Functionf("Reserving appNum %d for %s\n",
+			appNum, appNumOnUNetKey(baseID, appID))
+		baseMap.Set(appNum)
+		// Clear InUse
+		uuidpairtonum.AppNumFree(log, pubUUIDPairToNum, baseID, appID)
+	}
+
+	// In case zedrouter process restarted we fill in InUse from
+	// AppNetworkStatus, underlay network entries
+	items = pubAppNetworkStatus.GetAll()
+	for _, item := range items {
+		status := item.(types.AppNetworkStatus)
+		appID := status.UUIDandVersion.UUID
+
+		// If we have a config for the UUID we should mark it as
+		// allocated; otherwise mark it as reserved.
+		// XXX however, on startup we are not likely to have any
+		// config yet.
+		for i := range status.UnderlayNetworkList {
+			ulStatus := &status.UnderlayNetworkList[i]
+			baseID := ulStatus.Network
+			baseMap := appNumOnUNetBaseCreate(baseID)
+			appNum, err := uuidpairtonum.AppNumGet(log, pubUUIDPairToNum,
+				baseID, appID, appNumOnUNetType)
+			if err == nil {
+				if !baseMap.IsSet(appNum) {
+					log.Fatalf("Bitmap is not set for %s num %d",
+						appNumOnUNetKey(baseID, appID), appNum)
+				}
+				log.Functionf("(%s):Marking InUse for appNum %d\n",
+					appNumOnUNetKey(baseID, appID), appNum)
+				// Set InUse
+				uuidpairtonum.AppNumAllocate(log, pubUUIDPairToNum, baseID,
+					appID, appNum, false, appNumOnUNetType)
+			}
+		}
+	}
+}
+
+// If an entry is not inUse and and its CreateTime were
+// before the agent started, then we free it up.
+func appNumMapOnUNetGC(ctx *zedrouterContext) {
+
+	pub := ctx.pubUUIDPairToNum
+
+	log.Functionf("appNumOnUNetMapGC")
+	freedCount := 0
+	items := pub.GetAll()
+	for _, item := range items {
+		appNumMap := item.(types.UUIDPairToNum)
+		if appNumMap.NumType != appNumOnUNetType {
+			continue
+		}
+		if appNumMap.InUse {
+			continue
+		}
+		if appNumMap.CreateTime.After(ctx.agentStartTime) {
+			continue
+		}
+		appID := appNumMap.AppID
+		baseID := appNumMap.BaseID
+		log.Functionf("appNumMapOnUNetGC: freeing %s", appNumMap.Key())
+		appNumOnUNetFree(ctx, baseID, appID, true)
+		freedCount++
+	}
+	log.Functionf("appNumMapOnUNetGC freed %d", freedCount)
+}
+
+func appNumOnUNetAllocate(ctx *zedrouterContext, baseID uuid.UUID,
+	appID uuid.UUID, baseSize int, isZedmanager bool) int {
+	pub := ctx.pubUUIDPairToNum
+	baseMap := appNumOnUNetBaseCreate(baseID)
+
+	// Do we already have a number?
+	appNum, err := uuidpairtonum.AppNumGet(log, pub, baseID,
+		appID, appNumOnUNetType)
+	if err == nil {
+		log.Functionf("Found allocated appNum %d for %s\n", appNum,
+			appNumOnUNetKey(baseID, appID))
+		if appNum >= baseSize {
+			baseMap.Clear(appNum)
+			uuidpairtonum.AppNumDelete(log, ctx.pubUUIDPairToNum,
+				baseID, appID)
+			errStr := fmt.Sprintf(" app Num %d is outside %d",
+				appNum, baseSize)
+			log.Errorf("appNumOnUNetAllocate(%s): fail : %s",
+				appNumOnUNetKey(baseID, appID), errStr)
+			return types.AppNumInvalid
```

Comment:
Makes sense to have it return an error instead of this invalid because that isn't checked everywhere.
Hence function should return (int, error)